src/ray/gcs/gcs_server/gcs_resource_scheduler.cc

```diff
@@ -0,0 +1,301 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/gcs/gcs_server/gcs_resource_scheduler.h"
+
+namespace ray {
+namespace gcs {
+
+double LeastResourceScorer::Grade(const ResourceSet &required_resources,
+                                  const SchedulingResources &node_resources) {
+  const auto &available_resources = node_resources.GetAvailableResources();
+  const auto &available_resource_amount_map = available_resources.GetResourceAmountMap();
+
+  double node_score = 0.0;
+  for (const auto &entry : required_resources.GetResourceAmountMap()) {
+    auto available_resource_amount_iter = available_resource_amount_map.find(entry.first);
+    if (available_resource_amount_iter == available_resource_amount_map.end()) {
+      return -1;
+    }
+
+    auto calculate_score =
+        Calculate(entry.second, available_resource_amount_iter->second);
+    if (calculate_score < 0) {
+      return -1;
+    }
+    node_score += calculate_score;
+  }
+
+  // TODO(ffbin): When applying for CPU resources, if there are two nodes, one node has
+  // both GPU and CPU, and the other node only has CPU, we need to try our best to
+  // transfer to the node that only contains CPU. We will solve it in next pr.
+  return node_score;
+}
+
+double LeastResourceScorer::Calculate(const FractionalResourceQuantity &requested,
+                                      const FractionalResourceQuantity &available) {
+  if (available == 0 || requested > available) {
```

Comment:
Can we add `RAY_CHECK(available) >= 0`?  (or make sure it is always positive from FractionalResourceQuantity class?)

Reply:
FractionalResourceQuantity class can not make sure it is always positive, so i will add `RAY_CHECK(available) >= 0`.